addons/stock/models/product.py

```diff
@@ -739,14 +739,21 @@ def write(self, vals):
         if 'type' in vals and vals['type'] != 'product' and sum(self.mapped('nbr_reordering_rules')) != 0:
             raise UserError(_('You still have some active reordering rules on this product. Please archive or delete them first.'))
         if any('type' in vals and vals['type'] != prod_tmpl.type for prod_tmpl in self):
-            existing_move_lines = self.env['stock.move.line'].search([
+            existing_done_move_lines = self.env['stock.move.line'].search([
+                ('product_id', 'in', self.mapped('product_variant_ids').ids),
+                ('state', '=', 'done'),
+            ])
```

Comment:
Last comment about performance. As you want to trigger the error as soon as there is one move line, you can restrict the query with `limit=1`. Postgres will perform it faster.